tensorboard/webapp/plugins/npmi/views/annotations_list/annotation/annotation_component.ts

```diff
@@ -0,0 +1,662 @@
+    this.yScale = d3.scalePoint<string>().padding(0);
+    this.sizeScale = d3.scaleLinear().domain([0.0, 1.0]);
+    this.countSizeScale = d3.scaleLinear().range([2, 10]);
+    this.mainContainer = this.svg
+      .append('g')
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.barsGroup = this.mainContainer.append('g');
+    this.countDotsGroup = this.mainContainer.append('g');
+    this.textsGroup = this.mainContainer.append('g');
+    this.countTextsGroup = this.mainContainer.append('g');
+    this.runHintGroup = this.svg.append('g');
+    this.runClipPath = this.runHintGroup
+      .append('clipPath')
+      .attr('id', 'hint-clip')
+      .attr('class', 'hint-clip')
+      .append('rect')
+      .attr('x', 0)
+      .attr('y', 0)
+      .attr('width', this.margin.left - 30)
+      .attr('height', this.chartHeight);
+    this.redraw();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.svg) {
+      this.redraw();
+    }
+  }
+
+  private redraw() {
+    this.setTextClass();
+    this.updateAxes();
+    this.draw();
+  }
+
+  // Initializing/Updating the visualization props.
+  private setTextClass() {
+    this.textClass = 'default-text';
+    if (this.flaggedAnnotations.includes(this.annotation)) {
+      this.textClass = 'flag-text';
+    } else if (this.hiddenAnnotations.includes(this.annotation)) {
+      this.textClass = 'hidden-text';
+    }
+  }
+
+  private updateAxes() {
+    this.xScale
+      .rangeRound([
+        0,
+        this.chartWidth - this.chartWidth / this.activeMetrics.length,
+      ])
+      .domain(this.activeMetrics.map((d) => stripMetricString(d)));
+
+    this.yScale
+      .rangeRound([0, this.chartHeight - 30])
+      .domain(this.runs.map((d: string) => d));
+
+    this.sizeScale.range([0, this.chartWidth / this.activeMetrics.length]);
+
+    this.countSizeScale.domain([0, this.maxCount]);
+  }
+
+  // Drawing UI
+  private draw() {
+    this.drawRunIndicators();
+    this.drawRunHintTexts();
+    this.drawBars();
+    this.drawTexts();
+    if (this.showCounts) {
+      this.drawCountDots();
+      this.drawCountTexts();
+    } else {
+      this.countDotsGroup.selectAll('.count-dot').remove();
+      this.countTextsGroup.selectAll('.count-background-text').remove();
+      this.countTextsGroup.selectAll('.count-text').remove();
+    }
+  }
+
+  private drawRunIndicators() {
+    this.runClipPath.attr('height', this.chartHeight);
+
+    const indicators = this.runHintGroup
+      .selectAll('.hint')
+      .data([...this.runs]);
+
+    indicators
+      .enter()
+      .append('g')
+      .attr('class', 'hint')
+      .attr(
+        'transform',
+        function(this: AnnotationComponent, d: string) {
+          return `translate(10, ${this.yScale(d)! + 5})`;
+        }.bind(this)
+      )
+      .attr(
+        'fill',
+        function(this: AnnotationComponent, d: string) {
+          return `rgb(${this.rgbColors[0]})`;
+        }.bind(this)
+      )
```

Comment:
For things that has to be done on mutation (L234-L246), you can use `merge()`.

```ts
const indicatorEnters = indicators
      .enter()
      .append('g')
      .attr('class', 'hint');
indicatorEnters.append('path)...;

indicatorEnters.merge(indicators).attr('transform')...
```
If this approach works well, please apply it to `drawRunHintTexts` and others.